### PR TITLE
Refactor element access of pointer arrays into a separate function

### PR DESCRIPTION
**Pull request by antlechner**
~This PR is based on https://github.com/diffblue/cbmc/pull/4524. Please only review the last commit.~

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4b7cf97f30be8c030021b69a6e598f6e1098f40a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108164284

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 1]
```diff
@@ -191,6 +191,12 @@ dereference_exprt checked_dereference(const exprt &expr)
   return result;
 }
 
+dereference_exprt
+element_at_pointer_array_index(const exprt &pointer, const exprt &index)
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_utils.cpp]: Suggested name `array_element_from_pointer`

**Review by smowton — COMMENTED** [review]
This is not JBMC-specific -- suggest move to `util/`

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 2]
```diff
@@ -191,6 +191,12 @@ dereference_exprt checked_dereference(const exprt &expr)
...
+{
+  return dereference_exprt{plus_exprt{pointer, index}};
```
Comment (owen-mc-diffblue) [inline jbmc/src/java_bytecode/java_utils.cpp]: :bulb: Would it be a good idea to have something like `INVARIANT(can_cast_type<pointer_typet>(pointer.type()));` at the beginning of this function?

**Review by owen-mc-diffblue — APPROVED** [review]

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 2 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/java_utils.cpp]: Done (using `PRECONDITION`).

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 1 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/java_utils.cpp]: Changed the name to your suggestion.

**Comment by antlechner** [comment]
This is rebased on develop now.
@smowton I moved it to a new file `util/array_element_from_pointer.h` and changed the documentation slightly. I don't actually know how we represent arrays for C / when we represent them as pointers to the first element, so let me know if anything should be added to the documentation as it is now.

File: src/util/array_element_from_pointer.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+/// know the type of the first element of the array, so we work with that
+/// instead.
+/// \param pointer: pointer to the first element of an array
+/// \param index: index of the element to access
+/// \return expression representing the (\p index)'th element of the array
+dereference_exprt
+array_element_from_pointer(const exprt &pointer, const exprt &index)
+{
+  PRECONDITION(can_cast_type<pointer_typet>(pointer.type()));
+  return dereference_exprt{plus_exprt{pointer, index}};
```
Comment (tautschnig) [inline src/util/array_element_from_pointer.h:30]: While at it, maybe also do `PRECONDITION(index.type().id() == ID_signedbv || index.type().id() == ID_unsignedbv);`? I think those are the only two reasonable options, although maybe we could also work with some other bitvector types.

**Review by tautschnig — APPROVED** [review]
I think this is fine for C as well. I'm wondering, however, whether `util/pointer_offset_sum.h` would also be an ok header to place this? No particularly strong feelings about it, though.

File: src/util/array_element_from_pointer.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************\
+
+Module: Util
```
Comment (smowton) [inline src/util/array_element_from_pointer.h]: This is usually a description of the file, not the directory

File: src/util/array_element_from_pointer.h [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+#ifndef CPROVER_UTIL_ARRAY_ELEMENT_FROM_POINTER_H
+#define CPROVER_UTIL_ARRAY_ELEMENT_FROM_POINTER_H
+
+#include "std_expr.h"
+
+/// Generate statement using pointer arithmetic to access the element at the
+/// given index of a pointer array:
+/// `*(pointer + index)`
+/// Arrays are sometimes (always in JBMC) represented as a pointer to their
+/// first element. This is because we cannot know the type of an array of
```
Comment (smowton) [inline src/util/array_element_from_pointer.h]: Suggest replace the second sentence with just `, especially when their length is uncertain`

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 779f5dfb0e8ce4e44c451b59ebfca82b2e441009).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108340077

File: src/util/array_element_from_pointer.h (hunk 4 above)
Comment (antlechner) [reply src/util/array_element_from_pointer.h]: I had seen `Module: Util` in another file and just copied that. Most files in `util` seem to just leave this empty... Anyway, I changed this one to "Element access in a pointer array".

File: src/util/array_element_from_pointer.h (hunk 5 above)
Comment (antlechner) [reply src/util/array_element_from_pointer.h]: I replaced a part of the second sentence with that. :slightly_smiling_face:

File: src/util/array_element_from_pointer.h (hunk 3 above)
Comment (antlechner) [reply src/util/array_element_from_pointer.h:30]: I added the precondition. If we need other types in the future we can always add them later.

**Comment by antlechner** [comment]
@tautschnig `pointer_offset_sum.h` contains only one function with the same name as the file, so I'll keep this one separate for now to avoid confusion.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2088544ca4d07f73fbf60ab2b42f949f9bfac513).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108571279

**Pull request merged**